### PR TITLE
Simplify registration of custom functions in the parser and generator

### PR DESCRIPTION
**Pull request by izeigerman**
This is to simplify the extension of sqlglot with custom function.

 My assumption is that in a vast majority of cases generation of SQL expression from a custom function node is going to be straightforward and only in some cases a custom transformation is required.  

Thus I introduced a few improvements including a fallback logic in the Generator for functions for which a custom transformer hasn't been defined. 

The 1st minor simplification concerns registration of custom functions in the Parser:
```python
class SpecialUDF(Func):
    arg_types = {'a': True, 'b': False}

tokens = Tokenizer().tokenize('SELECT SPECIAL_UDF(a) FROM x')

expression = Parser(functions={
    **SpecialUDF.default_parser_mappings(),
    # **SpecialUDF2.default_parser_mappings()
    # etc
}).parse(tokens)[0]
```
Basically we try and infer the SQL function name from the class name.

The 2nd simplification is a bit more significant and concerns the Generator:
```python
expression.sql()
```
which returns:
```sql
SELECT SPECIAL_UDF(a) FROM x
```
In the simplest case users should no longer provide a custom transformer handler for their UDFs to generate the SQL expression.

### CODE REVIEW

**Comment by tobymao** [comment]
I'm a bit wary of this change, due to the performance implications of including  regex on every call. Also, why would you ever need to do provide a custom UDF without transformation? That's the only reason why you would do it. Any UDF can already be used without registration because it utilizes Anonymous function. Can you explain why you would need to pass in a custom UDF without transformation?

File: sqlglot/helper.py [hunk 1]
```diff
@@ -30,3 +31,8 @@ def list_get(arr, index):
 
 def csv(*args, sep=', '):
     return sep.join(arg for arg in args if arg)
+
+
+def camel_to_snake_case(name):
+    name = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)
```
Comment (tobymao) [inline sqlglot/helper.py:37]: if we were to have this, we should probably precompile the regex

Comment (tobymao) [reply sqlglot/helper.py:37]: if we do happen to do this change, which i'm not convinced of yet, i think we should require the user to provide the key of the expression so that it's static

**Comment by izeigerman** [comment]
@tobymao 
> UDF can already be used without registration because it utilizes Anonymous function.

This is an excellent point! 
There are, however, a few reasons to have a custom node without having a custom transformer:
1. Define proper argument keys to simplify tree traversal, transformation and argument extraction.
2. Extension of the previous point: benefit from the existing node validation by specifying concrete keys.
3. If I want to build a custom interpreter for the tree (not just SQL generation) it's so much easier to deal with properly typed nodes.

File: sqlglot/helper.py (hunk 1 above)
Comment (izeigerman) [reply sqlglot/helper.py:37]: Great point 👍 I didn't really consider performance implications.

**Comment by tobymao** [comment]
> @tobymao
> 
> > UDF can already be used without registration because it utilizes Anonymous function.
> 
> This is an excellent point!
> There are, however, a few reasons to have a custom node without having a custom transformer:
> 
> 1. Define proper argument keys to simplify tree traversal, transformation and argument extraction.
> 2. Extension of the previous point: benefit from the existing node validation by specifying concrete keys.
> 3. If I want to build a custom interpreter for the tree (not just SQL generation) it's so much easier to deal with properly typed nodes.

ok, that's fine, let's get rid of the dynamic camel_case in the hot loop

File: sqlglot/helper.py (hunk 1 above)
Comment (izeigerman) [reply sqlglot/helper.py:37]: Fixed this. Please have a look

Comment (tobymao) [reply sqlglot/helper.py:37]: can we precompile the regex?

File: sqlglot/generator.py [hunk 2]
```diff
@@ -538,6 +545,18 @@ def binary(self, expression, op, newline=False):
         sep = '\n' if newline else ' '
         return f"{self.sql(expression, 'this')}{sep}{op} {self.sql(expression, 'expression')}"
 
+    def function_fallback_sql(self, expression):
+        args = []
+        for arg_key in expression.arg_types.keys():
```
Comment (tobymao) [inline sqlglot/generator.py:550]: don't need keys

File: sqlglot/generator.py [hunk 3]
```diff
@@ -538,6 +545,18 @@ def binary(self, expression, op, newline=False):
...
+            arg_value = expression.args.get(arg_key) or []
+            if not isinstance(arg_value, list):
```
Comment (tobymao) [inline sqlglot/generator.py:552]: i feel like i've seen this pattern a bit, can we make a helper function called wrap_list(x: Any) -> List

**Comment by tobymao** [comment]
now that we have this, can we delete some unnecessary function lambdas in generator?

File: sqlglot/helper.py (hunk 1 above)
Comment (izeigerman) [reply sqlglot/helper.py]: Done that. Also decided to go with a simpler patter for now.

**Comment by izeigerman** [comment]
> now that we have this, can we delete some unnecessary function lambdas in generator?

@tobymao do you mean this piece - [link](https://github.com/tobymao/sqlglot/blob/main/sqlglot/generator.py#L427-L428) or something else?

File: sqlglot/generator.py (hunk 3 above)
Comment (izeigerman) [reply sqlglot/generator.py]: Added the `ensure_list` helper function and refactored accordingly.

**Review by tobymao — APPROVED** [review]
lgtm, were you able to prune some of the lambdas?

**Comment by tobymao** [comment]
> > now that we have this, can we delete some unnecessary function lambdas in generator?
> 
> @tobymao do you mean this piece - [link](https://github.com/tobymao/sqlglot/blob/main/sqlglot/generator.py#L427-L428) or something else?

https://github.com/tobymao/sqlglot/blob/main/sqlglot/generator.py#L59

**Comment by izeigerman** [comment]
@tobymao aaaah, got it! Sorry, misunderstood you the first time around. Will do this now. Guess I should also cleanup the dialects.py.

**Comment by izeigerman** [comment]
@tobymao done the clean up

**Comment by izeigerman** [comment]
Merging as per our discussion
